pkg/networkservice/common/timeout/server.go

```diff
@@ -32,69 +34,65 @@ import (
 )
 
 type timeoutServer struct {
-	onTimeout   *networkservice.NetworkServiceServer
 	connections map[string]*time.Timer
 	executor    serialize.Executor
 }
 
-// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections
-//             - onTimeout - *networkservice.NetworkServiceServer.  Since networkservice.NetworkServiceServer is an interface
-//                        (and thus a pointer) *networkservice.NetworkServiceServer is a double pointer.  Meaning it
-//                        points to a place that points to a place that implements networkservice.NetworkServiceServer
-//                        This is done because when we use timeout.NewServer as part of a chain, we may not *have*
-//                        a pointer to this server used 'onTimeout'.  If we detect we need to heal, onHeal.Request is used to heal.
-//                        If onTimeout is nil, then we simply set onTimeout to this server chain element
-//                        If we are part of a larger chain, we should pass the resulting chain into
-//                        this constructor before we actually have a pointer to it.
-func NewServer(onTimout *networkservice.NetworkServiceServer) networkservice.NetworkServiceServer {
-	rv := &timeoutServer{
+// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections.
+func NewServer() networkservice.NetworkServiceServer {
+	return &timeoutServer{
 		connections: make(map[string]*time.Timer),
-		executor:    serialize.NewExecutor(),
-		onTimeout:   onTimout,
-	}
-	if rv.onTimeout == nil {
-		var actualOnTimeout networkservice.NetworkServiceServer = rv
-		rv.onTimeout = &actualOnTimeout
 	}
-	return rv
 }
 
 func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	ct, err := t.createTimer(ctx, request)
+	conn, err := next.Server(ctx).Request(ctx, request)
 	if err != nil {
 		return nil, err
 	}
-	connID := request.GetConnection().GetId()
+
+	ct, err := t.createTimer(ctx, conn)
+	if err != nil {
+		if _, closeErr := next.Server(ctx).Close(ctx, conn); closeErr != nil {
+			return nil, errors.Wrapf(err, "Error attempting to close failed connection %v: %+v", conn.GetId(), closeErr)
+		}
+		return nil, err
+	}
+
+	connID := conn.GetId()
 	t.executor.AsyncExec(func() {
```

Comment:
I think we need to move timer.Stop before next.Server() stuff, since it is possible to close connection until it is refreshed.

Reply:
Done.